sopt/src/main/scala/dagr/sopt/cmdline/CommandLineParser.scala

```diff
@@ -0,0 +1,527 @@
+    *
+    * @param args              the command line arguments to parse.
+    * @param packageList       the list of packages to search for classes that extend [[Clp]].
+    * @param omitSubClassesOf  individual classes to omit from including on the command line.
+    * @param includeHidden     true to include classes whose [[clp]] annotation's hidden values is set to false.
+    * @param extraUsage        an optional string that will be printed prior to any usage message.
+    * @param clpBlock          a block of code to execute after a [[clp]] }instance has been created successfully.
+    * @return                  a [[clp]] instance initialized with the command line arguments, or [[None]] if
+    *                          unsuccessful.
+    * @return
+    */
+  private[cmdline] def parseClpHelper(args: Array[String],
+                                      packageList: List[String],
+                                      omitSubClassesOf: Iterable[Class[_]] = Nil,
+                                      includeHidden: Boolean = false,
+                                      extraUsage: Option[String] = None,
+                                      clpBlock: Clp => Unit.type = c => Unit): Option[Clp] = {
+
+    /** A brief developer note about the precedence of printing error messages and usages.
+      *
+      * See [[parseMainAndClp]] for 1-3, for when we have options to the main tool.
+      *
+      * 4. Complain if (a) no clp name is given, or (b) it is not recognized.
+      * 5. Complain if clp options are mis-specified.
+      * 6. Print the clp help message if --help is used after a clp name.
+      * 7. Print the version if --version is used after a clp name.
+      * 8. ... execute the clp ...
+      * */
+
+    val errorMessageBuilder: mutable.StringBuilder = new StringBuilder()
+
+    val print : (String => Unit) = System.err.println
+
+    def printExtraUsage(clazz: Option[Class[_]]): Unit = {
+      extraUsage match {
+        case Some(str) => print(str)
+        case None => print(standardClpUsagePreamble(clazz))
+      }
+    }
+
+    // Try parsing the task name
+    val classToClpAnnotation = findClpClasses[Clp](packageList, omitSubClassesOf = omitSubClassesOf, includeHidden = includeHidden)
+    parseClpName(args = args, classes = classToClpAnnotation.keySet) match {
+      case Right(error) => // Case 4 (b)
+        printExtraUsage(None)
+        print(clpListUsage(classToClpAnnotation.keySet, commandLineName, withPreamble=false))
+        print(wrapError(error))
+        None
+      case Left(clazz) =>
+        /////////////////////////////////////////////////////////////////////////
+        // Parse the arguments for the Command Line Program class
+        /////////////////////////////////////////////////////////////////////////
+        val clpParser = new CommandLineProgramParser(clazz)
+        clpParser.parseAndBuild(errorMessageBuilder, args.drop(1)) match {
```

Comment:
I settled on keeping the enumeration (since empty case classes are deprecated) but instead using a `Try` to indicate success or failure.  I also added a method to the `CommandLineProgramParser` object to perform all the usage printing in the case of help, version, or failure.